pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -0,0 +1,362 @@
+	old metav1.OwnerReference
+	new metav1.OwnerReference
+}
+
+// TODO: profile this function to see if a naive N^2 algorithm performs better
+// when the number of references is small.
+func referencesDiffs(old []metav1.OwnerReference, new []metav1.OwnerReference) (added []metav1.OwnerReference, removed []metav1.OwnerReference, changed []ownerRefPair) {
+	oldUIDToRef := make(map[string]metav1.OwnerReference)
+	for i := 0; i < len(old); i++ {
+		oldUIDToRef[string(old[i].UID)] = old[i]
+	}
+	oldUIDSet := sets.StringKeySet(oldUIDToRef)
+	newUIDToRef := make(map[string]metav1.OwnerReference)
+	for i := 0; i < len(new); i++ {
+		newUIDToRef[string(new[i].UID)] = new[i]
+	}
+	newUIDSet := sets.StringKeySet(newUIDToRef)
+
+	addedUID := newUIDSet.Difference(oldUIDSet)
+	removedUID := oldUIDSet.Difference(newUIDSet)
+	intersection := oldUIDSet.Intersection(newUIDSet)
+
+	for uid := range addedUID {
+		added = append(added, newUIDToRef[uid])
+	}
+	for uid := range removedUID {
+		removed = append(removed, oldUIDToRef[uid])
+	}
+	for uid := range intersection {
+		if !reflect.DeepEqual(oldUIDToRef[uid], newUIDToRef[uid]) {
+			changed = append(changed, ownerRefPair{old: oldUIDToRef[uid], new: newUIDToRef[uid]})
+		}
+	}
+	return added, removed, changed
+}
+
+// returns if the object in the event just transitions to "being deleted".
+func deletionStarts(e *event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so if there is no e.oldObj, we just return if the e.obj is being deleted.
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
+			return false
+		}
+	} else {
+		oldAccessor, err := meta.Accessor(e.oldObj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+			return false
+		}
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access newObj: %v", err))
+			return false
+		}
+		if accessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func beingDeleted(accessor meta.Object) bool {
+	return accessor.GetDeletionTimestamp() != nil
+}
+
+func hasDeleteDependentsFinalizer(accessor meta.Object) bool {
+	finalizers := accessor.GetFinalizers()
+	for _, finalizer := range finalizers {
+		if finalizer == v1.FinalizerDeleteDependents {
+			return true
+		}
+	}
+	return false
+}
+
+func hasOrphanFianlizer(accessor meta.Object) bool {
+	finalizers := accessor.GetFinalizers()
+	for _, finalizer := range finalizers {
+		if finalizer == v1.FinalizerOrphanDependents {
+			return true
+		}
+	}
+	return false
+}
+
+func startsWaitingForDependentsDeletion(e *event, accessor meta.Object) bool {
+	if !deletionStarts(e, accessor) {
+		// ignore the event if it's not about an object that starts being deleted.
+		return false
+	}
+	return hasDeleteDependentsFinalizer(accessor)
+}
+
+func shouldOrphanDependents(e *event, accessor meta.Object) bool {
```

Comment:
Renamed the function to "startsWaitingForDependentsOrphaned", is godoc still necessary?